File: aib-backend/src/main/java/com/aib/aib_backend/service/CategoryService.java
package com.aib.aib_backend.service;

import com.aib.aib_backend.dto.request.CategoryRequest;
import com.aib.aib_backend.dto.response.CategoryResponse;
import java.util.List;

/**
 * Service interface for Category operations
 */
public interface CategoryService {

    /**
     * Create a new category
     * @param request Category details
     * @return Created category
     */
    CategoryResponse createCategory(CategoryRequest request);

    /**
     * Update existing category
     * @param id Category ID
     * @param request Updated category details
     * @return Updated category
     */
    CategoryResponse updateCategory(Long id, CategoryRequest request);

    /**
     * Delete category (only if no products exist)
     * @param id Category ID
     */
    void deleteCategory(Long id);

    /**
     * Get category by ID
     * @param id Category ID
     * @return Category details with product count
     */
    CategoryResponse getCategoryById(Long id);

    /**
     * Get category by slug
     * @param slug Category slug
     * @return Category details with product count
     */
    CategoryResponse getCategoryBySlug(String slug);

    /**
     * Get all categories (admin view)
     * @return List of all categories with product counts
     */
    List<CategoryResponse> getAllCategories();

    /**
     * Get only active categories (public view)
     * @return List of active categories with product counts
     */
    List<CategoryResponse> getActiveCategories();
    CategoryResponse toggleCategoryStatus(Long id);
    int reactivateAllProductsInCategory(Long categoryId);
}

File: aib-frontend/src/app/core/services/category.service.ts
// category.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Category, CategoryRequest } from '@shared/models/category.model';
import { environment } from '@environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  private apiUrl = `${environment.apiUrl}/categories`;

  constructor(private http: HttpClient) {}

  // Public methods
  getActiveCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(this.apiUrl);
  }

  getCategoryById(id: number): Observable<Category> {
    return this.http.get<Category>(`${this.apiUrl}/${id}`);
  }

  getCategoryBySlug(slug: string): Observable<Category> {
    return this.http.get<Category>(`${this.apiUrl}/slug/${slug}`);
  }

  // Admin methods
  getAllCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(`${this.apiUrl}/admin/all`);
  }

  createCategory(request: CategoryRequest): Observable<Category> {
    return this.http.post<Category>(this.apiUrl, request);
  }

  updateCategory(id: number, request: CategoryRequest): Observable<Category> {
    return this.http.put<Category>(`${this.apiUrl}/${id}`, request);
  }

  toggleCategoryStatus(id: number): Observable<Category> {
    return this.http.patch<Category>(`${this.apiUrl}/${id}/toggle-status`, {});
  }

  deleteCategory(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
